File: composeApp/src/androidMain/kotlin/com/andercarotfg/weatherappkmp/weeklyWeather/WeeklyWeatherViewModel.kt
package ui.weeklyWeather

import com.db.WeatherAppDatabaseKMP
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.datetime.Clock
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import model.data.weatherBL
import model.db.DatabaseDriverFactory
import model.db.createDatabase
import model.db.actualWeather.actualWeatherDataSource
import model.db.actualWeather.actualWeatherRepositorySQL
import model.db.actualWeather.dailyWeatherRepositorySQL
import model.db.actualWeather.weeklyWeatherRepositorySQL
import model.db.dailyWeather.dailyWeatherDataSource
import model.db.weeklyWeather.weeklyWeatherDataSource
import model.domain.daySpecWeather

class WeeklyWeatherViewModel {

    val initListMaxMin = listOf(
        Pair(Triple("Sábado", "14", "9.3"), 0),
        Pair(Triple("Domingo", "14", "9.3"), 0),
        Pair(Triple("Lunes", "14", "9.3"), 0),
        Pair(Triple("Martes", "14", "9.3"), 0),
        Pair(Triple("Miercoles", "14", "9.3"), 0),
        Pair(Triple("Jueves", "14", "9.3"), 0),
        Pair(Triple("Viernes", "14", "9.3"), 0)
    )

    private val initListI = listOf(1, 2, 3, 2, 3, 1, 3)
    private val initListS = listOf("1", "2", "3", "4", "5", "6", "7")



    private val _latitude = MutableStateFlow<String>("43.5667")
    val latitude: StateFlow<String> = _latitude

    private val _longitude = MutableStateFlow<String>("-5.9")
    val longitude: StateFlow<String> = _longitude

    //private val _temperatures = MutableStateFlow<List<Double>>()
    //val temperatures: StateFlow<List<Double>> = _temperatures

    private val _maxMin = MutableStateFlow<List<Pair<Triple<String, String, String>, Int>>>(initListMaxMin)
    val maxMin: StateFlow<List<Pair<Triple<String, String, String>, Int>>> = _maxMin

    private val _codes = MutableStateFlow<List<Int>>(initListI)
    val codes: StateFlow<List<Int>> = _codes

    private val _hours = MutableStateFlow<List<String>>(initListS)
    val hours: StateFlow<List<String>> = _hours



    private val weekDayList: List<String> = listOf("Lunes", "Martes", "Miercoles", "Jueves",
        "Viernes", "Sábado", "Domingo", "Lunes", "Martes", "Miercoles", "Jueves", "Viernes",
        "Sábado")

    val db = createDatabase(DatabaseDriverFactory())
    val ds_ww = weeklyWeatherDataSource(db)
    val repo_ww = weeklyWeatherRepositorySQL(dataSource = ds_ww)
    suspend fun getAllData(latitude: Double, longitude: Double) {
        val weekW = weatherBL.getAllData(_latitude.value.toDouble(), _longitude.value.toDouble())
        val currentHour = Clock.System.now()
        val currentTime = currentHour.toLocalDateTime(TimeZone.UTC).dayOfWeek.ordinal
        val weekDaySpanish = weekDayList.get((currentTime + 1)-1)
        val res = repo_ww.getAll()

        res.collect { weeklyWeatherResult ->
            weeklyWeatherResult.onSuccess { ww ->
                var cambio = false
                if(ww.isNotEmpty()){
                    if (ww.first().longitude == _longitude.value.toDouble() || ww.first().latitude == _latitude.value.toDouble()) {
                        if (ww.first().date != weekDaySpanish) {
                            cambio = true
                            println("ANDER")
                        }
                    } else {
                        cambio = true
                        println("MIKEl")
                    }
                    var data: MutableList<Pair<Triple<String, String, String>, Int>> = mutableListOf()
                    for(i in ww){
                        data.add(Pair(Triple(i.date, i.temperatureMax.toString(), i.temperatureMin.toString()), i.code))
                    }
                    _maxMin.value = data

                }else{
                    cambio = true
                    println("NAIA")
                }
                if (cambio) {
                    repo_ww.deleteAll()
                    for (i in 1..7) {
                        val currentDay = weatherBL.getSpecificWeekDayTemperature(weekW, i)
                        val currentMaxMin = weatherBL.getMaxAndMinT(currentDay)
                        val mostRepeatedCode = weatherBL.getAverageCode(currentDay)
                        val weekDaySpanish = weekDayList.get((currentTime + i)-1)
                        repo_ww.insert(weekDaySpanish, _latitude.value.toDouble(),
                            _longitude.value.toDouble(),
                            currentMaxMin.first.toDouble(),
                            currentMaxMin.second.toDouble(), mostRepeatedCode.toLong())
                    }
                    var data: MutableList<Pair<Triple<String, String, String>, Int>> = mutableListOf()
                    for(i in ww){
                        data.add(Pair(Triple(i.date, i.temperatureMax.toString(), i.temperatureMin.toString()), i.code))
                    }
                    _maxMin.value = data
                }

            }.onFailure {

            }


        }
    }
    fun setLatAndLong(latitude: Double, longitude: Double) {
        _latitude.value = latitude.toString()
        _longitude.value = longitude.toString()
    }
}

File: iosApp/iosApp/ui/weeklyWeather/weeklyWeatherViewModel.swift
//
//  weeklyWeatherViewModel.swift
//  iosApp
//
//  Created by Ander Caro on 6/5/24.
//  Copyright © 2024 orgName. All rights reserved.
//
import Foundation
import Shared
import SwiftUI

class weeklyWeatherViewModel: ObservableObject{
    @Published var maxMin: [((String?, String?, String?), Int?)] = [
        (("Sábado", "14", "9.3"), 0),
        (("Domingo", "14", "9.3"), 0),
        (("Lunes", "14", "9.3"), 0),
        (("Martes", "14", "9.3"), 0),
        (("Miercoles", "14", "9.3"), 0),
        (("Jueves", "14", "9.3"), 0),
        (("Viernes", "14", "9.3"), 0)]
    
    @Published var latitude: Double = 0.0
    @Published var longitude:Double = 0.0
    
    private var WeatherBL: weatherBL = weatherBL()
    
    
    var db = DatabaseDriverFactoryKt.createDatabase(driverFactory: DatabaseDriverFactory())
    var cambio = false
    var hasRun = false
    var change = false
    
    func getAllData() async {
        print(latitude)
        print(longitude)
        print("Estoy en getAllData en iOS")
        do{
            let ds_ww = weeklyWeatherDataSource(db: self.db)
            let repo_ww = weeklyWeatherRepositorySQL(dataSource: ds_ww)
            
            try await repo_ww.deleteAll()
            
            try await repo_ww.getAlliOS().collect(collector: Collector<weeklyWeatherList?>{ value in
                Task{
                    do{
                        guard !self.hasRun else { return } // Asegura que solo se ejecute una vez
                        if(self.change){
                            self.hasRun = true
                        }
                        let weekW = try await self.WeatherBL.getAllData(latitude: self.latitude, longitude: self.longitude)
                        var lista: [((String?,String, String), Int)] = []
                        for i in 1...7{
                            let currentDay = self.WeatherBL.getSpecificWeekDayTemperature(weekWeather: weekW, dayNumber: Int32(i))
                            let currentMaxMin = self.getMaxAndMinT(dayWeather: currentDay)
                            let mostRepeatedCode = self.getAverageCode(dayWeather: currentDay)
                            let weekDaySpanish = self.getDayOfWeek(afterDays: i-1)?.capitalized
                            lista.append(((weekDaySpanish, currentMaxMin.0, currentMaxMin.1), mostRepeatedCode))
                        }
                        self.maxMin = lista
                        if(!value!.wekklyList.isEmpty){
                            if(value!.wekklyList.last?.latitude == self.latitude && value!.wekklyList.last?.longitude == self.longitude){
                                
                                if(value!.wekklyList.first?.date != lista.first?.0.0){
                                    self.cambio = true
                                }
                                else{
                                    print("weeklyWeather: VM -> BD -> VM")

                                    var listar: [((String?,String?, String?), Int?)] = []
                                    for i in 1...7{
                                        listar.append(((value!.wekklyList[i-1].date,
                                                        String(value!.wekklyList[i-1].temperatureMax),
                                                        String(value!.wekklyList[i-1].temperatureMin)),
                                                       Int(value!.wekklyList[i-1].code)))
                                    }
                                    self.maxMin = listar
                                }
                            }else{
                                self.cambio = true
                            }
                        }else{
                            self.cambio = true
                        }
                        if(self.cambio){
                            print("weeklyWeather: VM -> BD -> API -> BD -> VM")
                            try await repo_ww.deleteAll()
                            for i in 1...7{
                                try await repo_ww.insert(date: lista[i-1].0.0!, latitude: self.latitude, longitude: self.longitude, temperatureMax: Double(lista[i-1].0.1)!, temperatureMin: Double(lista[i-1].0.2)!, code: Int64(lista[i-1].1))
                            }
                            try await repo_ww.getAlliOS().collect(collector: Collector<weeklyWeatherList>{w in
                                var listar: [((String?,String?, String?), Int?)] = []
                                for i in 1...7{
                                    listar.append(((w.wekklyList[i-1].date,
                                                    String(w.wekklyList[i-1].temperatureMax),
                                                    String(w.wekklyList[i-1].temperatureMin)),
                                                   Int(w.wekklyList[i-1].code)))
                                }
                                self.maxMin = listar
                            })
                        }
                        
                       

                    }catch{
                        print(error)
                    }
                }
                
            })
            
        }
        catch {
                print(error)
            }
        
        }
    
    func setLatAndLong(Latitude: Double, Longitude: Double){
        self.change = true
           latitude = Latitude
           longitude = Longitude
       }
    
    func getDayOfWeek(afterDays days: Int) -> String? {
        let calendar = Calendar(identifier: .gregorian)
        let currentDate = Date()
        
        if let futureDate = calendar.date(byAdding: .day, value: days, to: currentDate) {
            let dateFormatter = DateFormatter()
            dateFormatter.locale = Locale(identifier: "es_ES")
            dateFormatter.dateFormat = "EEEE"
            return dateFormatter.string(from: futureDate)
        } else {
            return nil
        }
    }
    
    func getMaxAndMinT(dayWeather: dayWeather) -> (String, String) {
        var max = -1.0
        var min = 1000.0
            for t in dayWeather.temperatures {
                if (Double(truncating: t) > max) {
                    max = Double(truncating: t)
                } else if (Double(truncating: t)<min){
                    min = Double(truncating: t)
                }
            }
            return (String(max), String(min))
    }

    func getAverageCode(dayWeather: dayWeather) -> Int {
        
        let countedSet = NSCountedSet(array: dayWeather.codes)
        let mostFrequent = countedSet.max { countedSet.count(for: $0) < countedSet.count(for: $1) }
        return mostFrequent as! Int
    }

}

